dashboard/modules/job/job_supervisor.py

```diff
@@ -293,159 +950,173 @@ async def run(
             variables.
         3) Handle concurrent events of driver execution and
         """
-        curr_info = await self._job_info_client.get_info(self._job_id)
-        if curr_info is None:
-            raise RuntimeError(f"Status could not be retrieved for job {self._job_id}.")
-        curr_status = curr_info.status
-        curr_message = curr_info.message
-        if curr_status == JobStatus.RUNNING:
-            raise RuntimeError(
-                f"Job {self._job_id} is already in RUNNING state. "
-                f"JobSupervisor.run() should only be called once. "
-            )
-        if curr_status != JobStatus.PENDING:
-            raise RuntimeError(
-                f"Job {self._job_id} is not in PENDING state. "
-                f"Current status is {curr_status} with message {curr_message}."
-            )
-
-        if _start_signal_actor:
-            # Block in PENDING state until start signal received.
-            await _start_signal_actor.wait.remote()
-
-        driver_agent_http_address = (
-            "http://"
-            f"{ray.worker.global_worker.node.node_ip_address}:"
-            f"{ray.worker.global_worker.node.dashboard_agent_listen_port}"
-        )
-        driver_node_id = ray.worker.global_worker.current_node_id.hex()
-
-        await self._job_info_client.put_status(
-            self._job_id,
-            JobStatus.RUNNING,
-            jobinfo_replace_kwargs={
-                "driver_agent_http_address": driver_agent_http_address,
-                "driver_node_id": driver_node_id,
-            },
-        )
```

Comment:
All state management has been lifted up to `JobSupervisor` (to consolidate ownership scope, avoid race-conditions)